File: linkedlist/143.c
/**
 * 三步处理完, 先用快慢指针找出中点 之后reverse后半部分节点 最后merge到一起
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* myreverse(struct ListNode* head);
void reorderList(struct ListNode* head) {
    if(head == NULL || head->next == NULL) return;
    struct ListNode *pre, *fast, *slow, *l1;
    fast = head;
    slow = head;
    l1 = head;
    while(fast && fast->next)
    {
        pre = slow;
        slow = slow->next;
        fast = fast->next->next;
    }
    
    pre->next = NULL;
    struct ListNode* l2 = myreverse(slow);
    merge(l1,l2);
    
}

struct ListNode* myreverse(struct ListNode* head)
{
    struct ListNode *pre, *cur, *next;
    pre = NULL;
    cur = head;
    next = NULL;
    while(cur)
    {
        next = cur->next;
        cur->next = pre;
        pre = cur;
        cur = next;
    }
    return pre;
}

void merge(struct ListNode* l1, struct ListNode* l2)
{
    struct ListNode *t1, *t2;
    while(l1)
    {
        t1 = l1->next;
        t2 = l2->next;
        l1->next = l2;
        if(t1 == NULL) break;
        l2->next = t1;
        l1 = t1;
        l2 = t2;
    }
}

File: linkedlist/143.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null) {
            return;
        }
        int length = 0;
        ListNode cur = head;
        
        while (cur != null) {
            length++;
            cur = cur.next;
        }
        
        cur = head;
        int step = length / 2;
        while (step > 0) {
            cur = cur.next;
            step--;
        }
        ListNode nxt = cur.next;
        // 注意这里要断开
        cur.next = null;
        nxt = this.reverse(nxt);
        cur = head;
        while (nxt != null) {
            ListNode temp = cur.next;
            cur.next = nxt;
            nxt = nxt.next;
            cur.next.next = temp;
            cur = temp;
        }
        return;
        
        
    }
    
    public ListNode reverse(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode pre = dummy;
        ListNode next;
        while (head != null && head.next != null) {
            next = head.next;
            head.next = next.next;
            next.next = pre.next;
            pre.next = next;
        }
        return dummy.next;
    }
}
